oat-sa/extension-tao-testqti
# add test session mode

Test session object must not be acquired by two processes for writing because it may lead to inconsistency.

In order to prevent it new optional parameter `$forReadingOnly` added to `TestSessionService::getTestSession`:
```
\oat\taoQtiTest\models\TestSessionService::getTestSession(DeliveryExecution $deliveryExecution, $forReadingOnly = false);
```

1. Attempt to retrieve test session for *writing* by process B when it is already acquired for writing by process A should put process A in waiting more (see lock service) until lock is released.
2. Attempt to store session which retrieved for reading should lead to and exception.


**How to test:**
Put the following script into taoQtiTest/scripts/tools/:
https://gist.github.com/hutnikau/b739300cfef0128897f8a253f97121a4
Enable LockService:
sudo -u www-data php index.php '\oat\tao\scripts\install\SetUpLockService' --persistence redis

1. Run the test
2. Put a breakpoint on \oat\taoQtiTest\models\runner\synchronisation\action\Move::process():79
3. Press next button and make sure that process stopped at breakpoint
4. Run the script: sudo -u www-data php index.php '\oat\taoQtiTest\scripts\tools\CloseTestSession' <session_id>

Expected behavior: script should be waiting until breakpoint is released. After release of breakpoint the script should finish and close the session. Following pressing to the next button should fail with a message which says that test session is closed
